src/rust/jsg/wrappable.rs

```diff
@@ -0,0 +1,222 @@
+// Licensed under the Apache 2.0 license found in the LICENSE file or at:
+//     https://opensource.org/licenses/Apache-2.0
+
+//! Trait for converting between Rust and JavaScript values.
+//!
+//! # Supported Types
+//!
+//! | Rust Type | JavaScript Type |
+//! |-----------|-----------------|
+//! | `()` | `undefined` |
+//! | `String` | `string` |
+//! | `bool` | `boolean` |
+//! | `f64` | `number` |
+//! | `Option<T>` | `T` or `null` |
+//! | `Result<T, E>` | `T` or throws |
+//! | `NonCoercible<T>` | `T` |
+//! | `T: Struct` | `object` |
+
+use std::fmt::Display;
+
+use crate::Lock;
+use crate::NonCoercible;
+use crate::Struct;
+use crate::Type;
+use crate::v8;
+use crate::v8::ToLocalValue;
+
+impl Type for String {
+    type This = Self;
+
+    fn class_name() -> &'static str {
+        "string"
+    }
+
+    fn wrap<'a, 'b>(this: Self::This, lock: &'a mut Lock) -> v8::Local<'b, v8::Value>
+    where
+        'b: 'a,
+    {
+        this.to_local(lock)
+    }
+
+    fn is_exact(value: &v8::Local<v8::Value>) -> bool {
+        value.is_string()
+    }
+
+    fn unwrap(isolate: v8::IsolatePtr, value: v8::Local<v8::Value>) -> Self {
+        unsafe { v8::ffi::unwrap_string(isolate.as_ffi(), value.into_ffi()) }
+    }
+}
+
+impl Type for bool {
+    type This = Self;
+
+    fn class_name() -> &'static str {
+        "boolean"
+    }
+
+    fn wrap<'a, 'b>(this: Self::This, lock: &'a mut Lock) -> v8::Local<'b, v8::Value>
+    where
+        'b: 'a,
+    {
+        this.to_local(lock)
+    }
+
+    fn is_exact(value: &v8::Local<v8::Value>) -> bool {
+        value.is_boolean()
+    }
+
+    fn unwrap(isolate: v8::IsolatePtr, value: v8::Local<v8::Value>) -> Self {
+        unsafe { v8::ffi::unwrap_boolean(isolate.as_ffi(), value.into_ffi()) }
+    }
+}
+
+impl Type for f64 {
+    type This = Self;
+
+    fn class_name() -> &'static str {
+        "number"
+    }
+
+    fn wrap<'a, 'b>(this: Self::This, lock: &'a mut Lock) -> v8::Local<'b, v8::Value>
+    where
+        'b: 'a,
+    {
+        this.to_local(lock)
+    }
+
+    fn is_exact(value: &v8::Local<v8::Value>) -> bool {
+        value.is_number()
+    }
+
+    fn unwrap(isolate: v8::IsolatePtr, value: v8::Local<v8::Value>) -> Self {
+        unsafe { v8::ffi::unwrap_number(isolate.as_ffi(), value.into_ffi()) }
+    }
+}
+
+/// Trait for converting between Rust and JavaScript values.
+///
+/// Provides bidirectional conversion: `wrap` converts Rust to JavaScript,
+/// `unwrap` converts JavaScript to Rust. The `try_unwrap` method is used
+/// by macros to unwrap parameters with proper error handling.
+pub trait Wrappable: Sized {
+    /// Converts this Rust value into a JavaScript value.
+    fn wrap(self, lock: &mut Lock) -> v8::Local<'_, v8::Value>;
+
+    /// Converts a JavaScript value into this Rust type.
+    fn unwrap(isolate: v8::IsolatePtr, value: v8::Local<v8::Value>) -> Self;
+
+    /// Attempts to unwrap a JavaScript value, returning None and throwing a JS error on failure.
```

Comment:
what does `throwing a JS error` mean here in Rust context? panic? we shouldn't write code that is expected to panic.